### ISSUE TITLE
Functional Programming

### ISSUE THREAD

**Issue by knpwrs** [date-fns/date-fns#469]
This was mentioned in a comment previously: https://github.com/date-fns/date-fns/issues/275#issuecomment-266824617

I didn't see any followup responses, though, and the discussion as a whole seems like it's out of scope for #275. So here's a new issue!

The general idea to make the function signatures data-last (take `Date` objects, for instance, as their final argument) and auto-curry the functions to enable functional programming with libraries such as `ramda` or `lodash/fp`. Consider this brief, contrived example:

```js
const addOneDay = addDays(1);
const format = format('YYYY-MM-DD');
const formatNextDay = R.compose(format, addOneDay);
console.log(`Tomorrow's date is ${formatNextDay(new Date())}!`);
```

`date-fns` likens itself to be the "lodash for date functions" so maybe a `date-fns/fp` package would be appropriate, just like `lodash/fp`. I'm sure we'd all like to avoid [XKCD 927](https://xkcd.com/927/).

**Comment by leshakoss** [comment]
It's actually already implemented in [v2 branch](https://github.com/date-fns/date-fns/tree/v2). You can play with it:

```bash
git clone https://github.com/date-fns/date-fns.git date-fns-v2
cd date-fns-v2
git checkout v2
yarn
env PACKAGE_OUTPUT_PATH="$(pwd)/../YOUR-MODULE/node_modules/date-fns" ./scripts/buildPackage.sh
cd ..
```
(date-fns v2 will appear in `./YOUR-MODULE/node_modules/date-fns`)

Please note that this is pre-alpha and API of the library can and will be changed before release.

I think it works almost exactly as you would expect it. See doc: https://github.com/date-fns/date-fns/blob/v2/docs/fp.md

v2 will be released very soon, can't promise an exact date though.

(closed as duplicate of #253)

**Comment by knpwrs** [comment]
Derp! Sorry, I searched for `functional` and didn't even see #253.
